checkpoint_service.py

```diff
@@ -56,6 +71,72 @@ def _restore_from_checkpoint_file(self):
     def __del__(self):
         self._checkpoint_file_append_fp.close()
 
+
+class CheckpointKeyMap(AbstractCheckpointKeyMap):
+    """Deals with checkpoint read and write. Unlike CheckpointKeySet, it also saves the value.
+    Useful when the corresponding value is needed.
+    """
+    def __init__(self, checkpoint_file):
+        """
+        :param checkpoint_file: file to read / write object keys for checkpointing
+        """
+        self._checkpoint_file = checkpoint_file
+        self._checkpoint_key_map = {}
+        self._checkpoint_file_append_fp = ThreadSafeWriter(checkpoint_file, 'a')
+        self._restore_from_checkpoint_file()
+
+    def write(self, key, value):
+        if key not in self._checkpoint_key_map or "IN_USE_BY" in self._checkpoint_key_map[key]:
+            self._checkpoint_key_map[key] = value
+            self._checkpoint_file_append_fp.write(json.dumps({"key": str(key), "value": str(value)}) + "\n")
+
+    def check_contains_otherwise_mark_in_use(self, key):
+        """
+        If the key_map does not have the key value yet, mark the key to be IN_USE_BY_$THREAD_ID, and return False.
```

Comment:
Why do we mark as `IN_USE_BY_$THREAD_ID`? Is the expectation that the current thread will subsequently populate & set the value?

Reply:
That seems reasonable but might just be good to have a comment to clarify

Reply:
yes exactly. This is needed because of the behavior we desire which is to make one thread be responsible for a runs and all of its parent runs. In this behavior we can run into this situation where:
1. Thread A is assigned  run A whose parents are run E -> run T
2. Thread B is assigned run T

Case 1 is when thread A takes run T before Thread B gets there.
Thread A takes and mark the run T's value to be `IN_USE_BY_$THREAD_ID_OF_THREAD_A`
Thread B tries to take but sees that it is in Use. It has to wait until the real value is available because it has to return the run Id.
Thread A finishes the work, and correctly sets the value to be the new run Id.
Thread B sees the new value and return. 

Case 2 is when thread B takes run T before Thread A gets there.
Thread B takes and mark the run T's value to be `IN_USE_BY_$THREAD_ID_OF_THREAD_B`
Thread A tries to take but sees that it is in Use. It has to wait until the real value is available.
Thread B finishes the work and sets the value to be the the new run Id.
Thread A sees the new value and return **which is used as run E's parent run Id as the recursion returns.**

Reply:
let me leave the comment that the whoever is calling this to set, should also set the value later